### PR TITLE
Show all export statements on `brew shellenv` if users modified `PATH` explicitly

### PR DESCRIPTION
**Pull request by XuehaiPan**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [X] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

In recent days, users complain about the inconsistent shell experience since PR #11789 has been merged. Usually, the inconsistent behavior is caused by users modify the path variables (e.g. `PATH`) explicitly in their dotfiles.

This PR adds more documents for the variable `HOMEBREW_SHELLENV_PREFIX`. And adds warning to users if the variable `PATH` has been overridden in their dotfiles. Here, only `PATH` will be checked. `MANPATH` and `INFOPATH` will be ignored.

### CODE REVIEW

File: Library/Homebrew/brew.sh [hunk 1]
```diff
@@ -51,23 +51,8 @@ HOMEBREW_CACHE="${HOMEBREW_CACHE:-${HOMEBREW_DEFAULT_CACHE}}"
...
-  shellenv)            source "${HOMEBREW_LIBRARY}/Homebrew/cmd/shellenv.sh"; homebrew-shellenv; exit 0 ;;
-  formulae)            source "${HOMEBREW_LIBRARY}/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
-  casks)               source "${HOMEBREW_LIBRARY}/Homebrew/cmd/casks.sh"; homebrew-casks; exit 0 ;;
-  # falls back to cmd/prefix.rb on a non-zero return
-  --prefix*)           source "${HOMEBREW_LIBRARY}/Homebrew/prefix.sh"; homebrew-prefix "$@" && exit 0 ;;
-esac
-
 #####
-##### Next, define all helper functions.
+##### Next, define helper functions for prompting.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:55]:
```suggestion
##### Next, define output helper functions.
```

File: Library/Homebrew/cmd/shellenv.sh [hunk 2]
```diff
@@ -4,13 +4,24 @@
...
 # HOMEBREW_REPOSITORY is set by bin/brew
 # shellcheck disable=SC2154
 homebrew-shellenv() {
-  [[ "${HOMEBREW_SHELLENV_PREFIX}" == "${HOMEBREW_PREFIX}" ]] && return
+  if [[ "${HOMEBREW_SHELLENV_PREFIX}" == "${HOMEBREW_PREFIX}" ]]; then
+    [[ "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]] && return
+    opoo "You have set HOMEBREW_SHELLENV_PREFIX=\"${HOMEBREW_SHELLENV_PREFIX}\"," \
+         "but the brew's executable is not present in your PATH." \
+         "Please run \`brew help shellenv\` for more information."
+  fi
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/shellenv.sh:24]:
```suggestion
  if [[ "${HOMEBREW_SHELLENV_PREFIX}" == "${HOMEBREW_PREFIX}" && 
        "$(PATH="${HOMEBREW_PATH}" command -v brew)" == "${HOMEBREW_PREFIX}/bin/brew" ]] &&
        return
```

File: Library/Homebrew/cmd/shellenv.sh [hunk 3]
```diff
@@ -4,13 +4,24 @@
...
 #:  The variables `HOMEBREW_PREFIX`, `HOMEBREW_CELLAR` and `HOMEBREW_REPOSITORY` are also exported to avoid querying them multiple times.
 #:  The variable `HOMEBREW_SHELLENV_PREFIX` will be exported to avoid adding duplicate entries to the environment variables.
-#:  Consider adding evaluation of this command's output to your dotfiles (e.g. `~/.profile`, `~/.bash_profile`, or `~/.zprofile`) with: `eval $(brew shellenv)`
+#:
+#:  Consider adding evaluation of this command's output to your dotfiles (e.g. `~/.profile`, `~/.bash_profile`, or `~/.zprofile`)
+#:  with: `eval "$(/path/to/brew shellenv)"`.
+#:
+#:  The variable `HOMEBREW_SHELLENV_PREFIX` may be inherited from the parent shell session. If you explicitly
+#:  modified the path variables (e.g. `PATH`) in your dotfiles which breaks the subshell varibale inheritecement,
+#:  please use `eval "$(HOMEBREW_SHELLENV_PREFIX='' /path/to/brew shellenv)"` to always get all the export statements.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/shellenv.sh:13]: Can reset all of this.

File: Library/Homebrew/brew.sh [hunk 4]
```diff
@@ -125,6 +110,25 @@ odie() {
...
+  formulae)            source "${HOMEBREW_LIBRARY}/Homebrew/cmd/formulae.sh"; homebrew-formulae; exit 0 ;;
+  casks)               source "${HOMEBREW_LIBRARY}/Homebrew/cmd/casks.sh"; homebrew-casks; exit 0 ;;
+  # falls back to cmd/prefix.rb on a non-zero return
+  --prefix*)           source "${HOMEBREW_LIBRARY}/Homebrew/prefix.sh"; homebrew-prefix "$@" && exit 0 ;;
+esac
+
+#####
+##### Next, define the reset of helper functions.
+#####
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/brew.sh:131]: Can move this all back.

**Review by MikeMcQuaid — COMMENTED** [review]
I'd suggest instead if we're going to do a `command -v` anyway that we make that part of the early `return` and then we don't need to document the behaviour change.

**Review by carlocab — APPROVED** [review]
Thanks, @XuehaiPan!

**Comment by carlocab** [comment]
Let's try this out.

**Comment by MikeMcQuaid** [comment]
Thanks for this @XuehaiPan!

**Pull request merged**